causallearn/utils/cit.py

```diff
@@ -243,6 +240,40 @@ def _Fill3DCountTable(dataSXY, cardSXY):
         SyJointCounts = np.sum(SxyJointCounts, axis=1)
         return SxyJointCounts, SMarginalCounts, SxJointCounts, SyJointCounts
 
+    def _Fill3DCountTable_by_unique(dataSXY, cardSXY):
+        # Sometimes when the conditioning set contains many variables and each variable's cardinality is large
+        # e.g. consider an extreme case where
+        # S contains 7 variables and each's cardinality=20, then cardS = np.prod(cardSXY[:-2]) would be 1280000000
+        # i.e., there are 1280000000 different possible combinations of S,
+        #    so the SxyJointCounts array would be of size 1280000000 * cardX * cardY * np.int64,
+        #    i.e., ~3.73TB memory! (suppose cardX, cardX are also 20)
+        # However, samplesize is usually in 1k-100k scale, far less than cardS,
+        # i.e., not all (and actually only a very small portion of combinations of S appeared in data)
+        #    i.e., SMarginalCountsNonZero in _Fill3DCountTable_by_bincount is a very sparse array
+        # So when cardSXY is large, we first re-index S (skip the absent combinations) and then count XY table for each.
+        # See https://github.com/cmu-phil/causal-learn/pull/37.
+        cardX, cardY = cardSXY[-2:]
+        cardSs = cardSXY[:-2]
+
+        cardSsCumProd = np.ones_like(cardSs)
+        cardSsCumProd[:-1] = np.cumprod(cardSs[1:][::-1])[::-1]
+        SIndexed = np.dot(cardSsCumProd[None], dataSXY[:-2])[0]
+
+        uniqSIndices, inverseSIndices, SMarginalCounts = np.unique(SIndexed, return_counts=True, return_inverse=True)
+        cardS_reduced = len(uniqSIndices)
+        SxyIndexed = inverseSIndices * cardX * cardY + dataSXY[-2] * cardY + dataSXY[-1]
+        SxyJointCounts = np.bincount(SxyIndexed, minlength=cardS_reduced * cardX * cardY).reshape((cardS_reduced, cardX, cardY))
+
+        SxJointCounts = np.sum(SxyJointCounts, axis=2)
+        SyJointCounts = np.sum(SxyJointCounts, axis=1)
+        return SxyJointCounts, SMarginalCounts, SxJointCounts, SyJointCounts
+
+    def _Fill3DCountTable(dataSXY, cardSXY):
+        # about the threshold 1e5, see a rough performance example at:
+        # https://gist.github.com/MarkDana/e7d9663a26091585eb6882170108485e#file-count-unique-in-array-performance-md
+        if np.prod(cardSXY) < 1e5: return _Fill3DCountTable_by_bincount(dataSXY, cardSXY)
```

Comment:
nit: It would be better to define 1e5 as a constant (with meaningful naming). 

Numbers like this are called "magic number in code", and generally we should avoid that.

https://codeburst.io/software-anti-patterns-magic-numbers-7bc484f40544